### PR TITLE
Finalized validator index cache saving

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

part of https://github.com/prysmaticlabs/prysm/pull/13944 breakup

It may need to be removed or updated as part of ~~https://github.com/ethereum/consensus-specs/pull/3689~~ https://github.com/ethereum/consensus-specs/pull/3818

this PR adds saving to the validator index cache but is feature-gated, 
we will remove the feature gate when appropriate

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

File: beacon-chain/state/state-native/setters_misc.go [hunk 1]
```diff
@@ -114,6 +114,10 @@ func (b *BeaconState) SaveValidatorIndices() {
 		return
 	}
 
+	if b.validatorIndexCache == nil {
```
Comment (james-prysm) [inline beacon-chain/state/state-native/setters_misc.go:117]: does initializing the cache here make sense?

**Review by terencechain — APPROVED** [review]

File: beacon-chain/state/state-native/setters_misc.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/setters_misc.go:117]: Shouldn't that be during the lock?

File: beacon-chain/state/state-native/setters_misc.go [hunk 2]
```diff
@@ -114,6 +114,10 @@ func (b *BeaconState) SaveValidatorIndices() {
...
+		b.validatorIndexCache = newFinalizedValidatorIndexCache()
+	}
+
 	b.lock.Lock()
 	defer b.lock.Unlock()
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/setters_misc.go:122]:
```suggestion
	b.lock.Lock()
	defer b.lock.Unlock()
	
	if b.validatorIndexCache == nil {
		b.validatorIndexCache = newFinalizedValidatorIndexCache()
	}
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
One suggestion

File: beacon-chain/state/state-native/setters_misc.go (hunk 1 above)
Comment (james-prysm) [reply beacon-chain/state/state-native/setters_misc.go:117]: hmm that's true

Comment (james-prysm) [reply beacon-chain/state/state-native/setters_misc.go:120]: accepted suggestion *

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**